src/posts/2020-07-27-unsucking-your-project-onboarding.md

```diff
@@ -0,0 +1,789 @@
+   this is not mandatory (the script should handle this).
+3. We use `asdf` to manage versions, although this is not mandatory (the script
+   should also handle this).
+
+With this in mind, let's breakdown a simple `bin/setup` script. The first step
+is usually to ensure the needed languages are installed. Afterwards we should
+set up the environment variables and source them. Finally, install any needed
+dependencies and create and migrate the database.
+
+#### Helpers
+
+Before we got dive into the script, a small detour. I usually add a
+`bin/functions` script containing some helper functions. It gets passed around
+from project to project and all it does is provide some pretty printing
+functions, a `not_installed` helper and functions to `ensure_confirmation` and
+`ask_confirmation`:
+
+```shell
+#!/usr/bin/env sh
+
+BLUE='\033[1;34m'
+GREEN='\033[1;32m'
+YELLOW='\033[1;99m'
+RED='\033[1;91m'
+RESET='\033[0m'
+
+pp() {
+  printf "$1[$2]: $3${RESET}\n"
+}
+
+pp_info() {
+  pp $BLUE "$1" "$2"
+}
+
+pp_success() {
+  pp $GREEN "$1" "$2"
+}
+
+pp_error() {
+  pp $RED "$1" "$2"
+}
+
+pp_warn() {
+  pp $YELLOW "$1" "$2"
+}
+
+not_installed() {
+  [ ! -x "$(command -v "$@")" ]
+}
+
+ensure_confirmation() {
+  read -r "confirmation?please confirm you want to continue [y/n] (default: y) "
+  confirmation=${confirmation:-"y"}
+
+  if [ "$confirmation" != "y" ]; then
+    exit 1
+  fi
+}
+
+ask_confirmation() {
+  read -r "confirmation?please confirm you want to continue [y/n] (default: y) "
+  confirmation=${confirmation:-"y"}
+
+  [[ "$confirmation" == "y" ]];
+}
+```
+
+#### Language Install
+
+Having this, let's set up our script:
+
+```shell
+#!/usr/bin/env sh
+
+set -e
+. "./bin/functions"
+
+env=${1:-"dev"}
+```
+
+As usual we use the `-e` flag to make sure we stop the script if any command
+fails and we `source` the `bin/functions` helpers. We also allow defining an
+`env`, which defaults to `dev` as the first argument.
+
+The first step is then to install the required languages. In this sample
+project, let's assume we are using Elixir for the backend and the frontend is a
+classic React app.
+
+```shell
+pp_info "setup" "Installing required languages..."
+
+if not_installed "asdf"; then
+  pp_error "setup" "
+    We are using asdf (https://github.com/asdf-vm/asdf) to manage tool
+    dependencies, since it was not found on your system we cannot ensure that you
+    are using the correct versions of all the tools. Please install it and run
+    this script again, or proceed at your own peril.
+  "
+
+  ensure_confirmation
+else
+  set +e
+  asdf plugin-add erlang https://github.com/asdf-vm/asdf-erlang.git 2>/dev/null
+  asdf plugin-add elixir https://github.com/asdf-vm/asdf-elixir.git 2>/dev/null
+  asdf plugin-add nodejs https://github.com/asdf-vm/asdf-nodejs.git 2>/dev/null
+  set -e
+
+  printf "Importing nodejs release team keyring... "
+  bin/helpers/import-nodejs-keys
+  echo "Done!"
```

Comment:
I disagree with this, in the sense that I'd rather have a developer not having to deal with this and quickly join in than having to go through those battles themselves. I really disagree with the philosophy of "let's make it difficult just because, since that way they'll learn".

Either way, I'll reference the script provided by `asdf-node`